File: scripts/analyze-heap.ts
#!/usr/bin/env bun

interface HeapSnapshot {
  snapshot: {
    meta: any
    node_count: number
    edge_count: number
  }
  nodes: any[]
  edges: any[]
  strings: string[]
}

// Field indices for node structure
const NODE_TYPE_OFFSET = 0
const NODE_NAME_OFFSET = 1
const NODE_ID_OFFSET = 2
const NODE_SELF_SIZE_OFFSET = 3
const NODE_EDGE_COUNT_OFFSET = 4
const NODE_TRACE_NODE_ID_OFFSET = 5

// Field indices for edge structure
const _EDGE_TYPE_OFFSET = 0
const _EDGE_NAME_OR_INDEX_OFFSET = 1
const _EDGE_TO_NODE_OFFSET = 2

// Node types
const NODE_TYPE = [
  "hidden",
  "array",
  "string",
  "object",
  "code",
  "closure",
  "regexp",
  "number",
  "native",
  "synthetic",
  "concatenated string",
  "sliced string",
  "symbol",
  "bigint",
]

// Edge types
const _EDGE_TYPE = ["context", "element", "property", "internal", "hidden", "shortcut", "weak"]

function parseNode(nodes: any[], nodeIndex: number, strings: string[]) {
  const nodeFields = 6 // number of fields per node
  const nodeOffset = nodeIndex * nodeFields

  const type = nodes[NODE_TYPE_OFFSET][nodeOffset]
  const nameIndex = nodes[NODE_NAME_OFFSET][nodeOffset]
  const id = nodes[NODE_ID_OFFSET][nodeOffset]
  const selfSize = nodes[NODE_SELF_SIZE_OFFSET][nodeOffset]
  const edgeCount = nodes[NODE_EDGE_COUNT_OFFSET][nodeOffset]
  const traceNodeId = nodes[NODE_TRACE_NODE_ID_OFFSET]?.[nodeOffset]

  return {
    type: NODE_TYPE[type] || `type_${type}`,
    name: strings[nameIndex] || "",
    id,
    selfSize,
    edgeCount,
    traceNodeId,
    index: nodeIndex,
  }
}

function analyzeSnapshot(filePath: string) {
  console.log(`\n📊 Analyzing: ${filePath}`)

  const content = Bun.file(filePath).text()
  const snapshot: HeapSnapshot = JSON.parse(content)

  const { nodes, edges, strings } = snapshot
  const nodeArray = nodes
  const _edgeArray = edges

  // Parse all nodes and group by type
  const typeStats = new Map<string, { count: number; selfSize: number; retainedSize: number }>()
  const constructorStats = new Map<string, { count: number; selfSize: number }>()

  let totalNodes = 0
  let totalSelfSize = 0

  // Calculate number of nodes
  const _nodeFields = 6
  const numNodes = nodeArray[NODE_TYPE_OFFSET].length

  for (let i = 0; i < numNodes; i++) {
    const node = parseNode(nodeArray, i, strings)

    totalNodes++
    totalSelfSize += node.selfSize

    // Stats by node type
    if (!typeStats.has(node.type)) {
      typeStats.set(node.type, { count: 0, selfSize: 0, retainedSize: 0 })
    }
    const typeStat = typeStats.get(node.type)!
    typeStat.count++
    typeStat.selfSize += node.selfSize

    // Stats by constructor/name (filter out internal nodes)
    if (node.name && !node.name.startsWith("(")) {
      if (!constructorStats.has(node.name)) {
        constructorStats.set(node.name, { count: 0, selfSize: 0 })
      }
      const constStat = constructorStats.get(node.name)!
      constStat.count++
      constStat.selfSize += node.selfSize
    }
  }

  return {
    filePath,
    totalNodes,
    totalSelfSize: totalSelfSize / 1024 / 1024, // Convert to MB
    typeStats: Array.from(typeStats.entries())
      .map(([type, stats]) => ({
        type,
        count: stats.count,
        selfSizeMB: stats.selfSize / 1024 / 1024,
      }))
      .sort((a, b) => b.selfSizeMB - a.selfSizeMB),
    constructorStats: Array.from(constructorStats.entries())
      .map(([name, stats]) => ({
        name,
        count: stats.count,
        selfSizeMB: stats.selfSize / 1024 / 1024,
      }))
      .sort((a, b) => b.selfSizeMB - a.selfSizeMB),
  }
}

function compareSnapshots(snapshot1: any, snapshot2: any) {
  console.log("\n🔍 COMPARISON ANALYSIS")
  console.log("=".repeat(80))

  // Overall metrics
  console.log("\n📈 Overall Metrics:")
  console.log(
    `  Before: ${(snapshot1.totalSelfSize).toFixed(2)} MB (${snapshot1.totalNodes.toLocaleString()} nodes)`
  )
  console.log(
    `  After:  ${(snapshot2.totalSelfSize).toFixed(2)} MB (${snapshot2.totalNodes.toLocaleString()} nodes)`
  )
  console.log(
    `  Delta:  ${(snapshot2.totalSelfSize - snapshot1.totalSelfSize).toFixed(2)} MB (${(snapshot2.totalNodes - snapshot1.totalNodes).toLocaleString()} nodes)`
  )

  // Compare by type
  console.log("\n🏷️  Top Object Types by Growth:")
  console.log("─".repeat(80))
  console.log(`{'Type':<40} {'Before':>12} {'After':>12} {'Delta':>12} {'Delta Nodes':>15}`)
  console.log("─".repeat(80))

  const type1Map = new Map(snapshot1.typeStats.map((t) => [t.type, t]))
  const type2Map = new Map(snapshot2.typeStats.map((t) => [t.type, t]))

  const allTypes = new Set([
    ...snapshot1.typeStats.map((t) => t.type),
    ...snapshot2.typeStats.map((t) => t.type),
  ])

  const typeDeltas = Array.from(allTypes)
    .map((type) => {
      const stats1 = type1Map.get(type) || { count: 0, selfSizeMB: 0 }
      const stats2 = type2Map.get(type) || { count: 0, selfSizeMB: 0 }
      return {
        type,
        beforeSize: stats1.selfSizeMB,
        afterSize: stats2.selfSizeMB,
        deltaSize: stats2.selfSizeMB - stats1.selfSizeMB,
        beforeCount: stats1.count,
        afterCount: stats2.count,
        deltaCount: stats2.count - stats1.count,
      }
    })
    .sort((a, b) => b.deltaSize - a.deltaSize)

  for (const stat of typeDeltas.slice(0, 15)) {
    const deltaStr =
      stat.deltaSize >= 0 ? `+${stat.deltaSize.toFixed(2)}` : stat.deltaSize.toFixed(2)
    const countStr =
      stat.deltaCount >= 0
        ? `+${stat.deltaCount.toLocaleString()}`
        : stat.deltaCount.toLocaleString()
    console.log(
      `${stat.type.padEnd(40)} ${stat.beforeSize.toFixed(2).padStart(12)} MB ${stat.afterSize.toFixed(2).padStart(12)} MB ${deltaStr.padStart(12)} MB ${countStr.padStart(15)}`
    )
  }

  // Compare by constructor
  console.log("\n🏗️  Top Constructors by Growth:")
  console.log("─".repeat(80))
  console.log(`{'Constructor':<50} {'Before':>12} {'After':>12} {'Delta':>12} {'Delta Nodes':>15}`)
  console.log("─".repeat(80))

  const const1Map = new Map(snapshot1.constructorStats.map((c) => [c.name, c]))
  const const2Map = new Map(snapshot2.constructorStats.map((c) => [c.name, c]))

  const allConstructors = new Set([
    ...snapshot1.constructorStats.map((c) => c.name),
    ...snapshot2.constructorStats.map((c) => c.name),
  ])

  const constDeltas = Array.from(allConstructors)
    .map((name) => {
      const stats1 = const1Map.get(name) || { count: 0, selfSizeMB: 0 }
      const stats2 = const2Map.get(name) || { count: 0, selfSizeMB: 0 }
      return {
        name,
        beforeSize: stats1.selfSizeMB,
        afterSize: stats2.selfSizeMB,
        deltaSize: stats2.selfSizeMB - stats1.selfSizeMB,
        beforeCount: stats1.count,
        afterCount: stats2.count,
        deltaCount: stats2.count - stats1.count,
      }
    })
    .sort((a, b) => b.deltaSize - a.deltaSize)

  for (const stat of constDeltas.slice(0, 25)) {
    if (Math.abs(stat.deltaSize) < 0.01 && stat.deltaCount === 0) continue
    const deltaStr =
      stat.deltaSize >= 0 ? `+${stat.deltaSize.toFixed(2)}` : stat.deltaSize.toFixed(2)
    const countStr =
      stat.deltaCount >= 0
        ? `+${stat.deltaCount.toLocaleString()}`
        : stat.deltaCount.toLocaleString()
    console.log(
      `${stat.name.substring(0, 50).padEnd(50)} ${stat.beforeSize.toFixed(2).padStart(12)} MB ${stat.afterSize.toFixed(2).padStart(12)} MB ${deltaStr.padStart(12)} MB ${countStr.padStart(15)}`
    )
  }

  // Key findings
  console.log("\n🎯 KEY FINDINGS:")
  console.log("=".repeat(80))

  const significantGrowths = constDeltas.filter((d) => d.deltaSize > 0.5).slice(0, 10)
  if (significantGrowths.length > 0) {
    console.log("\n⚠️  Significant Memory Growth (>0.5 MB):")
    for (const stat of significantGrowths) {
      console.log(
        `  • ${stat.name}: +${stat.deltaSize.toFixed(2)} MB (${stat.afterCount.toLocaleString()} instances, +${stat.deltaCount.toLocaleString()})`
      )
    }
  }

  const highCountGrowths = constDeltas.filter((d) => d.deltaCount > 100).slice(0, 10)
  if (highCountGrowths.length > 0) {
    console.log("\n📊 High Instance Count Growth (>100 instances):")
    for (const stat of highCountGrowths) {
      console.log(
        `  • ${stat.name}: +${stat.deltaCount.toLocaleString()} instances (+${stat.deltaSize.toFixed(2)} MB)`
      )
    }
  }

  // Look for specific Next.js/React related patterns
  console.log("\n🔧 Next.js/React Related Objects:")
  const reactPatterns = constDeltas
    .filter(
      (d) =>
        d.name.includes("React") ||
        d.name.includes("Component") ||
        d.name.includes("Element") ||
        d.name.includes("Fiber") ||
        d.name.includes("Hook") ||
        d.name.includes("Module") ||
        d.name.includes("turbopack") ||
        d.name.includes("hmr") ||
        d.name.includes("closure")
    )
    .slice(0, 15)

  if (reactPatterns.length > 0) {
    for (const stat of reactPatterns) {
      if (Math.abs(stat.deltaSize) < 0.01 && stat.deltaCount === 0) continue
      const deltaStr =
        stat.deltaSize >= 0 ? `+${stat.deltaSize.toFixed(2)}` : stat.deltaSize.toFixed(2)
      const countStr =
        stat.deltaCount >= 0
          ? `+${stat.deltaCount.toLocaleString()}`
          : stat.deltaCount.toLocaleString()
      console.log(`  ${stat.name}: ${deltaStr} MB (${countStr} instances)`)
    }
  }
}

async function main() {
  const snapshot1Path = "/Users/ahmadabdullah/Desktop/Heap-20260103T155934.heapsnapshot"
  const snapshot2Path = "/Users/ahmadabdullah/Desktop/Heap-20260103T155955.heapsnapshot"

  console.log("🔬 HEAP SNAPSHOT ANALYSIS")
  console.log("=".repeat(80))

  const snapshot1 = analyzeSnapshot(snapshot1Path)
  const snapshot2 = analyzeSnapshot(snapshot2Path)

  compareSnapshots(snapshot1, snapshot2)

  console.log(`\n${"=".repeat(80)}`)
  console.log("✅ Analysis complete!")
}

main().catch(console.error)

File: scripts/analyze-heap.py
#!/usr/bin/env python3

import json
import sys

NODE_FIELD_COUNT = 6
NODE_TYPES = [
    'hidden', 'array', 'string', 'object', 'code', 'closure',
    'regexp', 'number', 'native', 'synthetic', 'concatenated string',
    'sliced string', 'symbol', 'bigint', 'object shape'
]

def load_snapshot(path):
    print(f"Loading {path.split('/')[-1]}...")
    with open(path) as f:
        data = json.load(f)
    return data

def analyze_snapshot(data):
    snapshot = data['snapshot']
    nodes = data['nodes']
    strings = data['strings']
    node_count = snapshot['node_count']

    type_stats = {}
    constructor_stats = {}
    total_self_size = 0

    for i in range(node_count):
        offset = i * NODE_FIELD_COUNT
        node_type = nodes[offset + 0]
        name_idx = nodes[offset + 1]
        self_size = nodes[offset + 3]

        type_name = NODE_TYPES[node_type] if node_type < len(NODE_TYPES) else f'type_{node_type}'
        name = strings[name_idx] if name_idx < len(strings) else ''

        total_self_size += self_size

        # Track by type
        if type_name not in type_stats:
            type_stats[type_name] = {'count': 0, 'self_size': 0}
        type_stats[type_name]['count'] += 1
        type_stats[type_name]['self_size'] += self_size

        # Track by constructor name
        if name and not name.startswith('(') and not name.startswith('<') and len(name) > 0:
            if name not in constructor_stats:
                constructor_stats[name] = {'count': 0, 'self_size': 0}
            constructor_stats[name]['count'] += 1
            constructor_stats[name]['self_size'] += self_size

    return {
        'total_nodes': node_count,
        'total_self_size_mb': total_self_size / 1024 / 1024,
        'type_stats': type_stats,
        'constructor_stats': constructor_stats,
    }

def main():
    before_file = '/Users/ahmadabdullah/Desktop/Heap-20260103T162216.heapsnapshot'
    after_file = '/Users/ahmadabdullah/Desktop/Heap-20260103T162240.heapsnapshot'

    print('🔬 HEAP SNAPSHOT COMPARISON')
    print('=' * 100)
    print(f"\n📁 BEFORE: {before_file.split('/')[-1]}")
    print(f"📁 AFTER:  {after_file.split('/')[-1]}\n")

    before_data = load_snapshot(before_file)
    after_data = load_snapshot(after_file)

    before_stats = analyze_snapshot(before_data)
    after_stats = analyze_snapshot(after_data)

    # Overall metrics
    print('\n📊 OVERALL METRICS')
    print('─' * 100)
    print(f"{'Metric':<25} {'Before':>20} {'After':>20} {'Delta':>20}")
    print('─' * 100)

    size_delta = after_stats['total_self_size_mb'] - before_stats['total_self_size_mb']
    count_delta = after_stats['total_nodes'] - before_stats['total_nodes']

    print(f"{'Total Heap Size':<25} {before_stats['total_self_size_mb']:>20.2f} MB {after_stats['total_self_size_mb']:>20.2f} MB {size_delta:>+20.2f} MB")
    print(f"{'Total Node Count':<25} {before_stats['total_nodes']:>20,} {after_stats['total_nodes']:>20,} {count_delta:>+20,}")

    # Compare by type
    print('\n\n🏷️  MEMORY BY OBJECT TYPE (Top 20 by Delta)')
    print('─' * 100)

    all_types = set(before_stats['type_stats'].keys()) | set(after_stats['type_stats'].keys())

    type_deltas = []
    for t in all_types:
        before = before_stats['type_stats'].get(t, {'count': 0, 'self_size': 0})
        after = after_stats['type_stats'].get(t, {'count': 0, 'self_size': 0})
        delta = (after['self_size'] - before['self_size']) / 1024 / 1024
        type_deltas.append({
            'type': t,
            'before_size': before['self_size'] / 1024 / 1024,
            'after_size': after['self_size'] / 1024 / 1024,
            'delta_size': delta,
            'before_count': before['count'],
            'after_count': after['count'],
            'delta_count': after['count'] - before['count'],
        })

    type_deltas.sort(key=lambda x: abs(x['delta_size']), reverse=True)

    print(f"{'Type':<30} {'Before':>12} {'After':>12} {'Delta MB':>12} {'Delta Count':>12}")
    print('─' * 100)

    for stat in type_deltas[:20]:
        if abs(stat['delta_size']) < 0.01:
            continue
        delta_str = f"+{stat['delta_size']:.2f}" if stat['delta_size'] >= 0 else f"{stat['delta_size']:.2f}"
        count_str = f"+{stat['delta_count']:,}" if stat['delta_count'] >= 0 else f"{stat['delta_count']:,}"
        print(f"{stat['type'][:30]:<30} {stat['before_size']:>12.2f} MB {stat['after_size']:>12.2f} MB {delta_str:>12} MB {count_str:>12}")

    # Compare by constructor
    print('\n\n🏗️  TOP CONSTRUCTORS BY MEMORY GROWTH')
    print('─' * 100)

    all_constructors = set(before_stats['constructor_stats'].keys()) | set(after_stats['constructor_stats'].keys())

    constructor_deltas = []
    for name in all_constructors:
        before = before_stats['constructor_stats'].get(name, {'count': 0, 'self_size': 0})
        after = after_stats['constructor_stats'].get(name, {'count': 0, 'self_size': 0})
        delta = (after['self_size'] - before['self_size']) / 1024 / 1024
        constructor_deltas.append({
            'name': name,
            'before_size': before['self_size'] / 1024 / 1024,
            'after_size': after['self_size'] / 1024 / 1024,
            'delta_size': delta,
            'before_count': before['count'],
            'after_count': after['count'],
            'delta_count': after['count'] - before['count'],
        })

    constructor_deltas = [d for d in constructor_deltas if d['after_count'] > 0 and abs(d['delta_size']) > 0.01]
    constructor_deltas.sort(key=lambda x: x['delta_size'], reverse=True)

    print(f"{'Constructor':<55} {'Before':>10} {'After':>10} {'Delta':>10} {'Total':>8}")
    print('─' * 100)

    for stat in constructor_deltas[:30]:
        delta_str = f"+{stat['delta_size']:.2f}" if stat['delta_size'] >= 0 else f"{stat['delta_size']:.2f}"
        print(f"{stat['name'][:55]:<55} {stat['before_size']:>10.2f} MB {stat['after_size']:>10.2f} MB {delta_str:>10} MB {stat['after_count']:>8,}")

    # By instance count growth
    print('\n\n📊 TOP CONSTRUCTORS BY INSTANCE GROWTH')
    print('─' * 100)

    by_instance = sorted(constructor_deltas, key=lambda x: x['delta_count'], reverse=True)
    by_instance = [d for d in by_instance if d['delta_count'] > 0]

    print(f"{'Constructor':<55} {'Before':>10} {'After':>10} {'Delta':>10}")
    print('─' * 100)

    for stat in by_instance[:25]:
        count_str = f"+{stat['delta_count']:,}"
        print(f"{stat['name'][:55]:<55} {stat['before_count']:>10,} {stat['after_count']:>10,} {count_str:>10}")

    # Framework-specific
    print('\n\n🔧 NEXT.JS / REACT / TURBOPACK OBJECTS')
    print('─' * 100)

    patterns = ['turbopack', 'react', 'next', 'router', 'module', 'closure', 'fiber', 'component', 'hmr']
    framework_objs = [d for d in constructor_deltas if any(p in d['name'].lower() for p in patterns)]

    if framework_objs:
        print(f"{'Constructor':<55} {'Delta MB':>12} {'Delta Count':>12}")
        print('─' * 100)
        for stat in framework_objs[:20]:
            delta_str = f"+{stat['delta_size']:.2f}" if stat['delta_size'] >= 0 else f"{stat['delta_size']:.2f}"
            count_str = f"+{stat['delta_count']:,}" if stat['delta_count'] >= 0 else f"{stat['delta_count']:,}"
            print(f"{stat['name'][:55]:<55} {delta_str:>12} MB {count_str:>12}")
    else:
        print('  (No significant framework object growth detected)')

    # Key findings
    print('\n\n🎯 KEY FINDINGS')
    print('=' * 100)

    big_growth = [d for d in constructor_deltas if d['delta_size'] > 0.1]
    if big_growth:
        print('\n⚠️  Significant memory growth (>0.1 MB):')
        for stat in big_growth[:10]:
            print(f"  • {stat['name']}: +{stat['delta_size']:.2f} MB ({stat['after_count']:,} instances)")
    else:
        print('\n✅ No significant memory growth detected!')

    print('\n' + '=' * 100)

if __name__ == '__main__':
    main()
